### PR TITLE
feat: add p2p URL support across all content in the reader

### LINKED ISSUE #17: Support for Loading of p2p URLs for All Content
Support for p2p URLs within the content, such as images, videos, and logos, is lacking. This issue aims to implement functionality to detect p2p URLs everywhere and utilize gateways when necessary to ensure seamless loading of all p2p content in Social Reader.

### PR DESCRIPTION
**Pull request by akhileshthite**
closes #17 

Facing two issues:

1. Getting null src for `img` and `video` in the post content. The src attributes might be empty or not set at the time when the script tries to update them?

```javascript
    const images = contentDOM.querySelectorAll('img')
    images.forEach(img => {
      const src = img.getAttribute('src')
      console.log('Original src:', src)
      img.setAttribute('src', resolveP2PUrl(src))
    })

    const videos = contentDOM.querySelectorAll('video source')
    videos.forEach(video => {
      const src = video.getAttribute('src')
      video.setAttribute('src', resolveP2PUrl(src))
    })
```

2. With the same updates in random sort `timelines` branch, getting this error:
Screenshot 2024-05-26 at 3 13 29 AM

### CODE REVIEW

**Review by RangerMauve — CHANGES_REQUESTED** [review]
We shouldn't be resolving to gateways in browsers that support p2p urls. Instead see if there's a DOM event that gets emitted when an image/video fails to load and listen to it at the `document.body` then take the `event.target` and attempt to rewrite its `src`. Video takes also have different ways of specifying sets of sources so we should account for that too.

**Comment by RangerMauve** [comment]
For the `searchNotes` bit, check that you fixed the issues in #13 where the function got accidentally deleted

**Comment by akhileshthite** [comment]
> We shouldn't be resolving to gateways in browsers that support p2p urls. Instead see if there's a DOM event that gets emitted when an image/video fails to load and listen to it at the `document.body` then take the `event.target` and attempt to rewrite its `src`. Video takes also have different ways of specifying sets of sources so we should account for that too.

Yep, will add the fallback logic.

This approach is not working:

```javascript
    // Setup error listeners for media elements within this component
    this.addEventListener('error', event => {
      const target = event.target;
      if (target.tagName === 'IMG' || (target.tagName === 'SOURCE' && target.parentElement.tagName === 'VIDEO')) {
        const originalSrc = target.getAttribute('src');
        if (!originalSrc.startsWith('hyper://') && !originalSrc.startsWith('ipns://')) {
          target.setAttribute('src', resolveP2PUrl(originalSrc));
        }
      }
    }, true);
  }
```

**Comment by RangerMauve** [comment]
What part doesn't work? What steps have you taken to debug it so far?

**Comment by akhileshthite** [comment]
> What part doesn't work? What steps have you taken to debug it so far?

I experimented with load events just to verify the functionality and discovered  that IPNS and HYPER URLs are resolving as `null`, including the `original/initial` URLs. It may be due to the specific timing of how we handle p2p URLs, particularly when elements are processed before they are fully loaded or accessible in the DOM?

Example code:-
```javascript
  connectedCallback() {
      // Listen for load events on the entire document body for images and video sources
      document.body.addEventListener('load', event => {
        const target = event.target;
    
        // Check if the loaded element is an image or a video source
        if (target.tagName === 'IMG' || (target.tagName === 'SOURCE' && target.closest('video'))) {
          const originalSrc = target.getAttribute('src');
    
          if (originalSrc && !originalSrc.startsWith('hyper://') && !originalSrc.startsWith('ipns://')) {
            const newSrc = resolveP2PUrl(originalSrc);
            console.log('Rewriting src for:', target, 'from:', originalSrc, 'to:', newSrc);
            target.setAttribute('src', newSrc);
          }
        }
      }, true);
  }
```

**Comment by RangerMauve** [comment]
By resolving as null do you mean the `originalSrc` attribute is showing up as null or is the `resolveP2PURL` function returning null?

**Comment by RangerMauve** [comment]
```
if (!originalSrc.startsWith('hyper://') && !originalSrc.startsWith('ipns://')) {
          target.setAttribute('src', resolveP2PUrl(originalSrc));
        }
```

This seems to only be trying to resolve the URL  if it *isn't* p2p. Should the if statement be resolved and have us try to update the `src` attribute right away?

Which browser are you testing this on? Which URL/post?

**Comment by RangerMauve** [comment]
Mind committing your code?

**Comment by akhileshthite** [comment]
> Mind committing your code?

Fixed it!
Will commit in a few hours.

**Comment by akhileshthite** [comment]
> > Mind committing your code?
> 
> Fixed it! Will commit in a few hours.

Resolving urls before rendering in `loadAndRenderPost` worked: https://github.com/hyphacoop/reader.distributed.press/pull/20/commits/675329fbd6051b2bb9d1b02200000842a771a962

Regarding the gateway fallback how is this approach in `db.js`?
```javascript
export async function supportsP2P (url) {
  console.log(url)
  try {
    const response = await fetch(url)
    return response.ok
  } catch (error) {
    console.log('P2P URL loading failed:', error)
    return false
  }
}

export async function resolveP2PUrl (url) {
  if (!url) return url

  console.log('Attempting to resolve P2P URL:', url)

  if (isP2P(url)) {
    const p2pSupported = await supportsP2P(url)

    if (!p2pSupported) {
      if (url.startsWith(HYPER_PREFIX)) {
        return url.replace(HYPER_PREFIX, 'https://hyper.hypha.coop/hyper/')
      } else if (url.startsWith(IPNS_PREFIX)) {
        return url.replace(IPNS_PREFIX, 'https://ipfs.hypha.coop/ipns/')
      }
    }
  }

  return url
}
```

**Comment by RangerMauve** [comment]
Can you elaborate on what the purpose of the proposed changes to db.js are?

File: actor-mini-profile.js [hunk 1]
```diff
@@ -50,7 +51,7 @@ class ActorMiniProfile extends HTMLElement {
     // Actor icon
     const img = document.createElement('img')
     img.className = 'profile-mini-icon'
-    img.src = iconUrl
+    img.src = resolveP2PUrl(iconUrl)
```
Comment (RangerMauve) [inline actor-mini-profile.js]: We shouldn't resolve to http unless there's an error loading. Maybe we should have a `p2p-image` custom element made which will attempt to render an `img` and fall back with the onerror handler? Then we can reuse the logic across pages.

File: db.js [hunk 2]
```diff
@@ -103,6 +115,8 @@ export class ActivityPubDB extends EventTarget {
     if (url && typeof url === 'object') {
       return url
     }
+    url = resolveP2PUrl(url)
```
Comment (RangerMauve) [inline db.js]: We shouldn't be trying to resolve to the gateway right away. Keep the current logic which resolve to the gateway only when the p2p logic fails.

File: post.js [hunk 3]
```diff
@@ -39,6 +40,48 @@ function timeSince (dateString) {
...
+        console.log(`Resolving to gateway URL due to error: ${fallbackSrc}`)
+        video.setAttribute('src', fallbackSrc)
+      }
+    }
+
+    video.setAttribute('src', p2pSrc)
+    console.log(`Set video src to: ${p2pSrc}`)
+  })
+
+  return contentDOM.body.innerHTML // Return the modified HTML to be inserted
```
Comment (RangerMauve) [inline post.js]: Do the onerror events actually fire here? I think we need to add these listeners *after* we've sanitized and rendered the html for the post.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

**Comment by akhileshthite** [comment]
> Can you elaborate on what the purpose of the proposed changes to db.js are?

If the fetch fails, it will resolve the p2p URLs using gateways.

**Comment by akhileshthite** [comment]
https://github.com/hyphacoop/reader.distributed.press/pull/20/commits/58366b88f0bbfa35b0515d5c904b71afc30e835f

The `p2p-media.js` is resolving to gateway if the fetch fails:
Screenshot 2024-06-13 at 7 41 12 PM

The problem is it's not replacing the src:
original `img` src:
Screenshot 2024-06-13 at 7 41 57 PM

`p2p-image` src:
Screenshot 2024-06-13 at 7 42 09 PM

File: p2p-media.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      this.handleError(source)
+      return // Skip setting the source if not supported
+    }
+
+    source.onerror = () => this.handleError(source)
+    this.video.appendChild(source)
+  }
+
+  handleError (source) {
+    const fallbackSrc = resolveP2PUrl(source.src)
```
Comment (RangerMauve) [inline p2p-media.js]: Instead, check if the current `img.src` is p2p

File: profile.html [hunk 5]
```diff
@@ -40,6 +40,7 @@
 <script type="module" src="./sidebar.js"></script>
 <script type="module" src="./actor-profile.js"></script>
 <script type="module" src="./post.js"></script>
+<script type="module" src="./p2p-media.js"></script>
```
Comment (RangerMauve) [inline profile.html]: We should have post.js import any of it's dependencies directly instead of doing it top level.

Similarly outbox should import post.

**Review by RangerMauve — CHANGES_REQUESTED** [review]
We should modify the `supportsP2P` function so it only checks for the p2p schemes once.

something like this

- have a global map of `<protocol promise<boolean> | boolean>`
- supports will get the scheme from the src
- check if the map has a value for the scheme `if(supportMap.has(scheme)) return await supportMap.get(scheme)`
- if doesn't have, set a promise for attempting to resolve the url scheme
- if request fails on a network error (no http response) set it to false and return false
- else set to true

This will reduce the number of http requests and improve performance.

**Comment by RangerMauve** [comment]
We already have a fallback for detecting when p2p fetch fails: https://github.com/hyphacoop/reader.distributed.press/blob/main/db.js#L116 Lets stick to that without adding new stuff

**Comment by akhileshthite** [comment]
`p2p-image` and `p2p-video` are working fine everywhere; the profile pictures in `actor-profile`, `actor-profile-mini`, and `followed-accounts`.

The media is not loading in the post content, I think there's something wrong with `insertImagesAndVideos`:
```
post.js:55 Updated p2p-image src after fallback: https://hyper.hypha.coop/hyper/hypha.coop/assets/images/posts/2022-05-12-ppl-sitting-unsplash.jpg
post.js:59 Replaced img with p2p-image having initial src: hyper://hypha.coop/assets/images/posts/2022-05-12-ppl-sitting-unsplash.jpg
```

I'm currently debugging the issue.

**Comment by akhileshthite** [comment]
> We should modify the `supportsP2P` function so it only checks for the p2p schemes once.
> 
> something like this
> 
> * have a global map of `<protocol promise<boolean> | boolean>`
> * supports will get the scheme from the src
> * check if the map has a value for the scheme `if(supportMap.has(scheme)) return await supportMap.get(scheme)`
> * if doesn't have, set a promise for attempting to resolve the url scheme
> * if request fails on a network error (no http response) set it to false and return false
> * else set to true
> 
> This will reduce the number of http requests and improve performance.

https://github.com/hyphacoop/reader.distributed.press/blob/b5f63705addda30211318df4daf1988e66882aa7/db.js#L48

File: p2p-media.js [hunk 6]
```diff
@@ -1,28 +1,43 @@
...
     super()
+    this.attachShadow({ mode: 'open' })
 
+    // Create an img element
     this.img = document.createElement('img')
-    this.attachShadow({ mode: 'open' })
-    this.shadowRoot.appendChild(this.img)
+
+    // Load the CSS file
+    const style = document.createElement('style')
```
Comment (RangerMauve) [inline p2p-media.js]: Couldn't we just add a regular style tag with `@mport` in the template?

TBH I think we should avoid the shadow root entirely and just embed the image into the plain dom and use regular css targetting the custom element.

Comment (RangerMauve) [reply p2p-media.js]: Same for the video tag

File: post.js [hunk 7]
```diff
@@ -40,42 +40,41 @@ function timeSince (dateString) {
...
 
-  // Replace all <img> tags with <p2p-image> tags
-  contentDOM.querySelectorAll('img').forEach(img => {
+  const imgPromises = Array.from(contentDOM.querySelectorAll('img')).map(async (img) => {
     const originalSrc = img.getAttribute('src')
-    console.log(`Original img src: ${originalSrc}`)
     const p2pImg = document.createElement('p2p-image')
     p2pImg.setAttribute('src', originalSrc)
+
+    p2pImg.addEventListener('load-error', (event) => {
```
Comment (RangerMauve) [inline post.js]: Do we need this here? Feels like we could encapsulate this logic entirely in the p2p image component.

File: post.js [hunk 8]
```diff
@@ -40,42 +40,41 @@ function timeSince (dateString) {
...
+    p2pVideo.addEventListener('load-error', (event) => {
+      console.log(`Error loading video at ${originalSrc}, fallback to ${event.detail.fallbackSrc}`)
+      p2pVideo.setAttribute('src', event.detail.fallbackSrc)
+      console.log('Updated p2p-video src after fallback:', p2pVideo.getAttribute('src'))
     })
+
     video.parentNode.replaceChild(p2pVideo, video)
   })
 
+  await Promise.all([...imgPromises, ...videoPromises])
```
Comment (RangerMauve) [inline post.js]: These no longer need to be promises. We can just do for loops over the list of sources and append the new elements.

File: post.js [hunk 9]
```diff
@@ -98,7 +97,7 @@ class DistributedPost extends HTMLElement {
     try {
       const content = await db.getNote(postUrl)
       if (content && content.content) {
-        content.content = insertImagesAndVideos(content.content) // Resolve URLs before rendering
+        content.content = await insertImagesAndVideos(content.content) // Resolve URLs before rendering
```
Comment (RangerMauve) [inline post.js]: We should remove the async and just convert the img/video tags in a regular function since the load happens in the component.

**Review by RangerMauve — CHANGES_REQUESTED** [review]

**Comment by sutty-coop** [comment]
@akhileshthite have you been able to apply th changes requested here?

**Comment by akhileshthite** [comment]
> @akhileshthite have you been able to apply th changes requested here?

On it
Thanks

**Comment by akhileshthite** [comment]
Now we can load the p2p content in non p2p browsers :)

Screenshot 2024-09-03 at 7 35 12 AM
Screenshot 2024-09-03 at 7 36 53 AM

**Review by RangerMauve — APPROVED** [review]
Looks good, great job! Merge it in whenever

**Pull request merged**